### DISCUSSION TITLE
contributions to/future of pymeasure?

### DISCUSSION THREAD

**Discussion by gatling-nrl** [pymeasure/pymeasure#638] [Q&A]
We've begun developing new experiment/data acquisition code in python, where previously we used LabVIEW. For decades NI has been a curator of instrument drivers and has quite a long list of supported instruments. Python does not seem to have one central home for instrument drivers... pymeasure being the closest I have found.

Therefore, we're contemplating basing our new code around pymeasure, but of course you don't have all our instruments already supported. Which is the first of my questions: what is the process for adding support? Do you welcome pull requests with code for previously unsupported instruments? Who ensures quality of the code in these PR?

The other question I had was the roadmap for pymeasure. Do you intend to grow this package and userbase? Is there a "steering committee" or how will you make decisions about significant feature additions or backwards-breaking changes? How stable do you intend for this code base to be?

Thanks!

**Comment by bilderbuchi** [comment]
Welcome, and thank you for your interest!

> Python does not seem to have one central home for instrument drivers... pymeasure being the closest I have found.

Indeed, there is lots of fragmentation :-(. You can check out a [rough catalog](https://pymeasure.github.io/python-lab-automation-landscape/catalog/) of relevant packages.

> what is the process for adding support? 

Have you studied the [docs](https://pymeasure.readthedocs.io/en/latest/) already?
See [this page ](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html)for how to create new instruments. See [this page](https://pymeasure.readthedocs.io/en/latest/dev/contribute.html) for the contributing guide.

> Do you welcome pull requests with code for previously unsupported instruments? Who ensures quality of the code in these PR?

Yes, of course, see [here](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html).
As you can tell from the list of open PRs, reviewing and merging added instruments takes effort. We are three active spare-time maintainers (one currently finishing his PhD), so it often take us longer than I like to review PRs, as we also have to balance that with important features/structural improvements.

> The other question I had was the roadmap for pymeasure.

Pymeasure has been "stable" and regularly used (as evidenced by the regular PRs with incoming instrument driver contributions) for a good number of years now. As we are already bottlenecked in maintainer capacity, we have no plans for specific activities around growing our userbase.
There is a [board](https://github.com/pymeasure/pymeasure/projects/9) with things we want to address before we call it "1.0" (but people have been using it for their experiment setups/projects all the time). Even now, we rarely (if ever?) break existing functionality.

> Is there a "steering committee" 

Hey, we are not Numpy. ;-)
We discuss in issues to flesh out design questions for new features. So far we three maintainers have managed to come to a consensus. Sometimes we pull in the original author (who is not active anymore). Sometimes we prototype different approaches in PRs, then choose one.

> How stable do you intend for this code base to be?

I mean, it's been effectively "stable" (as in, generally usable/useful) for a long time? There are a couple of things that maybe could break existing code (see above, to get in before 1.0), I intend to manage that with deprecation warnings.
